src/sage/algebras/down_up_algebra.py

```diff
@@ -0,0 +1,841 @@
+"""
+Down-Up Algebras
+
+AUTHORS:
+
+- Travis Scrimshaw (2023-4): initial version
+"""
+
+#*****************************************************************************
+#  Copyright (C) 2023 Travis Scrimshaw <tcscrims at gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  http://www.gnu.org/licenses/
+#*****************************************************************************
+
+from sage.misc.cachefunc import cached_method
+from sage.categories.algebras import Algebras
+from sage.categories.modules import Modules
+from sage.categories.rings import Rings
+from sage.combinat.free_module import CombinatorialFreeModule
+from sage.sets.non_negative_integers import NonNegativeIntegers
+from sage.categories.sets_cat import cartesian_product
+from sage.sets.family import Family
+from sage.misc.lazy_list import lazy_list
+from sage.misc.misc_c import prod
+from sage.modules.free_module import FreeModule
+
+class DownUpAlgebra(CombinatorialFreeModule):
+    r"""
+    The down-up algebra.
+
+    Let `R` be a commutative ring, and let `\alpha, \beta, \gamma \in R`.
+    The *down-up algebra* is the associative unital algebra
+    `DU(\alpha, \beta, \gamma)` generated by `d, u` with relations
+
+    .. MATH::
+
+        \begin{aligned}
+        d^2u & = \alpha dud + \beta ud^2 + \gamma d,
+        \\ du^2 & = \alpha udu + \beta u^2d + \gamma u.
+        \end{aligned}
+
+    The down-up algebra has a PBW-type basis given by
+
+    .. MATH::
+
+        \{ u^i (du)^j d^k \mid i,j,k \in \ZZ_{\geq 0} \}.
+
+    This algebra originates in the study of posets. For a poset `P`,
+    we define operators acting on `R[P]` by
+
+    .. MATH::
+
+        d(y) = \sum_x x \qquad\qquad u(y) = \sum_z z,
+
+    where `y` covers `x` and `z` covers `y`. For `r`-differential posets
+    we have `du - ud = r 1` and afford a representation of a
+    :class:`Weyl algebra <sage.algebras.weyl_algebra.DifferentialWeylAlgebra>`.
+    This is obtained from `DU(0, 1, 2r)`. For a `(q,r)`-differential poset,
+    we have the `d` and `u` operators satisfying
+
+    .. MATH::
+
+        \begin{aligned}
+        d^2u & = q(q+1) dud - q^3 ud^2 + r d,
+        \\ du^2 & = q(q+1) udu - q^3 u^2d + r u,
+        \end{aligned}
+
+    or `\alpha = q(q+1)`, `\beta = -q^3`, and `\gamma = r`. Specializing
```

Comment:
replacing "or" with "Put differently, we set" (i.e., starting a new sentence and making it clear, that we are just saying the same thing differently) would make it possibly easier to read.  (Another possibility:
```
    For a `(q,r)`-differential poset,
    we have `\alpha = q(q+1)`, `\beta = -q^3`, and `\gamma = r`, or, explicitly

    .. MATH::

        \begin{aligned}
        d^2u & = q(q+1) dud - q^3 ud^2 + r d,
        \\ du^2 & = q(q+1) udu - q^3 u^2d + r u.
        \end{aligned}
```

Reply:
Will change.

> Another thing: are you saying that the Weyl relation is satisfied in $DU(0,1,2)$?

Yes, and you can explicitly see that in the doctest on Young's lattice. The proof of this was given in Benkart-Roby.

Reply:
Are you sure it's not the other way round?  I can see that if I have $du-ud=rI$ I have a down-up algebra $DU(0,1,2)$, but not conversely.

(sorry I cannot check properly right now, I didn't even get to the other ticket yet)

Reply:
I thought I copied it correctly from Benkart-Roby. Assuming I did, then there might be a minor misprint in the paper (swapping what `d` and `u` mean in the Weyl algebra). I also did not check carefully. I will do so tomorrow morning.

(No problem; I am going to write my longer response now, must later than the "tomorrow" I had promised...)Perhaps a slightly better phrasing would be "...and it affords a representation..."

Reply:
Indeed, there was a problem. The Weyl algebra is supposed to only be a quotient of $DU(0, 1, 2)$. I have updated the doc accordingly.